File: paleo_workbench/ui/pages/preparation_page.py
from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import QHBoxLayout, QMessageBox, QSplitter, QVBoxLayout, QWidget

from paleo_workbench.ui import tokens
from paleo_workbench.ui.pages.boundary_panel import BoundaryPanel
from geoviz import CancellationToken

from paleo_workbench.ui.pages.factor_prepare_worker import FactorPrepareWorker
from paleo_workbench.ui.pages.contour_draft_worker import (
    ContourDraftResult,
    ContourDraftWorker,
    commit_contour_drafts,
)
from paleo_workbench.ui.owned_worker_job import OwnedWorkerJob
from paleo_workbench.ui.pages.factor_preview_grid import FactorPreviewGrid
from paleo_workbench.ui.pages.factor_task_panel import FactorTaskPanel
from paleo_workbench.ui.pages.well_table_panel import WellTablePanel
from paleo_workbench.workflow.factor_prepare_scheduler import (
    FactorPrepareBatchResult,
    FactorPrepareProgress,
    build_prepare_snapshot,
    commit_prepare_batch_result,
)
from paleo_workbench.workflow.well_qc import qc_summary, run_well_table_qc
from paleo_workbench.workflow.well_table import (
    attach_well_table_to_factor_task,
    sample_points_from_well_table,
    well_table_from_factor_task,
)

# Back-compat for tests that imported FactorPrepareResult from the page module.
FactorPrepareResult = FactorPrepareBatchResult


class PreparationPage(QWidget):
    """制备 page: tasks, WellTable, preview grid, boundary + async interpolation."""

    # Emitted after staged prepare results are committed to project.factor_map_tasks
    factor_maps_updated = Signal()
    # Emitted after ContourDraft generation mutates contour_drafts / maps
    contour_drafts_updated = Signal()
    generate_requested = Signal(str)  # method — app may handle when no project bound

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("PreparationPage")
        self._project = None
        self._tasks: list = []
        self._prepare_generation = 0
        self._prepare_job = OwnedWorkerJob(self)
        self._prepare_job.released.connect(self._clear_prepare_job)
        self._contour_job = OwnedWorkerJob(self)
        self._contour_job.released.connect(self._clear_contour_job)

        outer = QVBoxLayout(self)
        outer.setContentsMargins(
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
        )
        outer.setSpacing(tokens.SPACE_4)

        content = QHBoxLayout()
        content.setSpacing(tokens.SPACE_4)

        self.task_panel = FactorTaskPanel()
        content.addWidget(self.task_panel, 0)

        center = QSplitter(Qt.Orientation.Vertical)
        self.preview_grid = FactorPreviewGrid()
        self.well_table_panel = WellTablePanel()
        center.addWidget(self.preview_grid)
        center.addWidget(self.well_table_panel)
        center.setStretchFactor(0, 3)
        center.setStretchFactor(1, 2)
        content.addWidget(center, 1)

        self.boundary_panel = BoundaryPanel()
        content.addWidget(self.boundary_panel, 0)

        outer.addLayout(content, 1)

        self.task_panel.generate_requested.connect(self._on_generate_requested)
        self.task_panel.contour_draft_requested.connect(self._on_contour_draft_requested)
        self.well_table_panel.run_qc_btn.clicked.connect(self._on_run_well_qc)

    def set_project(self, project) -> None:
        """Bind the live ProjectDocument so batch generate can mutate factor_map_tasks."""
        # Project switch supersedes any in-flight prepare generation.
        if project is not self._project and self.is_prepare_running():
            self._prepare_generation += 1
            self._prepare_job.cancel()
        self._project = project
        self._refresh_well_table_view()

    def update_state(self, tasks: list) -> None:
        self._tasks = list(tasks or [])
        self.task_panel.update_state(self._tasks)
        self.preview_grid.update_state(self._tasks)
        self._refresh_well_table_view()

    def is_prepare_running(self) -> bool:
        return self._prepare_job.is_running

    def shutdown_workers(self, wait_ms: int = 3_000) -> bool:
        """Quit in-flight prepare thread (call before shell deleteLater)."""
        contour_joined = self._contour_job.shutdown(wait_ms)
        prepare_joined = self._prepare_job.shutdown(wait_ms)
        self._set_generate_enabled(True)
        return contour_joined and prepare_joined

    def _shutdown_contour_worker(self, wait_ms: int) -> None:
        self._contour_job.shutdown(wait_ms)
        if not self.is_prepare_running():
            self._set_generate_enabled(True)

    def is_contour_running(self) -> bool:
        return self._contour_job.is_running

    def _set_generate_enabled(self, enabled: bool) -> None:
        self.task_panel.generate_btn.setEnabled(enabled)
        self.task_panel.contour_draft_btn.setEnabled(enabled)
        self.well_table_panel.run_qc_btn.setEnabled(enabled)

    def _refresh_well_table_view(self) -> None:
        table = self._resolve_display_well_table()
        self.well_table_panel.update_from_well_table(table)

    def _resolve_display_well_table(self):
        """Prefer project.well_tables; else derive from first factor task samples."""
        if self._project is None:
            return None
        tables = getattr(self._project, "well_tables", None) or []
        if tables:
            return tables[0]
        tasks = list(getattr(self._project, "factor_map_tasks", None) or self._tasks or [])
        for task in tasks:
            params = getattr(task, "parameters", None) or {}
            points = params.get("sample_points") if isinstance(params, dict) else None
            if points:
                return well_table_from_factor_task(task)
        return None

    def _on_run_well_qc(self) -> None:
        if self._project is None:
            QMessageBox.information(self, "井点 QC", "请先打开或绑定工程。")
            return
        try:
            self._run_well_qc_impl()
        except Exception as exc:
            QMessageBox.warning(
                self,
                "井点 QC 失败",
                f"{exc.__class__.__name__}: {exc}",
            )

    def _run_well_qc_impl(self) -> None:
        table = None
        if self._project.well_tables:
            table = self._project.well_tables[0]
        elif self._project.factor_map_tasks:
            task = self._project.factor_map_tasks[0]
            table = well_table_from_factor_task(task)
            attach_well_table_to_factor_task(self._project, table, task)
        if table is None or not table.rows:
            QMessageBox.information(self, "井点 QC", "没有可检测的井点数据。")
            return
        run_well_table_qc(table)
        # Keep first table in project
        if not self._project.well_tables:
            self._project.well_tables.append(table)
        else:
            self._project.well_tables[0] = table
        # Sync cleaned sample_points back onto linked tasks
        for task in self._project.factor_map_tasks:
            if task.well_table_id == table.id or task is self._project.factor_map_tasks[0]:
                params = dict(task.parameters or {})
                params["sample_points"] = sample_points_from_well_table(table)
                task.parameters = params
                task.well_table_id = table.id
        self._refresh_well_table_view()
        summary = qc_summary(table)
        QMessageBox.information(
            self,
            "井点 QC",
            f"完成：ok={summary.get('ok', 0)} outlier={summary.get('outlier', 0)} "
            f"invalid_ratio={summary.get('invalid_ratio', 0)} missing={summary.get('missing', 0)}",
        )

    def _on_generate_requested(self, method: str) -> None:
        method = method or self.task_panel.selected_method()
        if self._project is None:
            self.generate_requested.emit(method)
            return
        if self.is_prepare_running():
            QMessageBox.information(self, "单因素图", "正在生成中，请稍候…")
            return
        self._start_prepare_worker(method)

    def _start_prepare_worker(self, method: str) -> None:
        self._set_generate_enabled(False)
        self._prepare_generation += 1
        generation = self._prepare_generation
        token = CancellationToken()
        # Snapshot on the host thread so scientific inputs match Stage-4 fingerprints.
        snapshot = build_prepare_snapshot(
            self._project,
            generation=generation,
            method=method,
        )
        self.task_panel.summary_label.setText(
            f"制备中… 任务 {len(snapshot.tasks)} · 生成代 {generation}"
        )
        worker = FactorPrepareWorker(
            self._project,
            method=method,
            cancellation_token=token,
            generation=generation,
            snapshot=snapshot,
        )
        self._prepare_job.start(
            worker,
            terminal_signals=(worker.finished, worker.failed, worker.cancelled),
            result_connections=(
                (worker.completed, self._on_prepare_completed),
                (worker.failed, self._on_prepare_failed),
                (worker.progress, self._on_prepare_progress),
                (worker.cancelled, self._on_prepare_cancelled),
            ),
            cancel=token.cancel,
            target=self._project,
        )

    def _clear_prepare_job(self) -> None:
        if not self.is_contour_running():
            self._set_generate_enabled(True)

    def _on_prepare_progress(self, update: FactorPrepareProgress) -> None:
        if update.generation != self._prepare_generation:
            return
        if self._prepare_job.target is not self._project:
            return
        msg = update.message or update.phase
        self.task_panel.summary_label.setText(
            f"制备中：复用 {update.clean} · 需计算 {update.dirty} · "
            f"已完成 {update.completed}/{update.total_tasks}"
            + (f" · {msg}" if msg else "")
        )

    def _on_prepare_completed(self, result: FactorPrepareBatchResult) -> None:
        target = self._prepare_job.target
        if target is None or self._project is not target:
            return
        if int(result.generation) != int(self._prepare_generation):
            # Superseded by a newer prepare request or project switch.
            return
        discarded = commit_prepare_batch_result(
            target,
            result,
            expected_generation=self._prepare_generation,
        )
        self.update_state(target.factor_map_tasks)
        self.factor_maps_updated.emit()
        extra = ""
        if discarded:
            extra = f" · 丢弃过期 {len(discarded)}"
        self.task_panel.summary_label.setText(
            f"已制备 {sum(1 for t in target.factor_map_tasks if t.status == 'complete')} / "
            f"{len(target.factor_map_tasks)} 个单因素图"
            f" · 复用 {result.clean_count} · 计算 {result.executed_count}{extra}"
        )

    def _on_prepare_failed(self, message: str) -> None:
        if self._prepare_job.target is not self._project:
            return
        # Async failures must not enter a nested modal loop while the shell may
        # be rebuilding. Keep the error visible and recoverable in-page.
        self.task_panel.summary_label.setText(f"单因素图生成失败：{message}")

    def _on_prepare_cancelled(self) -> None:
        if self._prepare_job.target is not self._project:
            return
        self.task_panel.summary_label.setText("单因素图生成已取消")

    def _on_contour_draft_requested(self) -> None:
        """Schedule ContourDraft extraction outside the GUI thread."""
        if self._project is None:
            QMessageBox.information(self, "等值线初稿", "请先打开或绑定工程。")
            return
        if self.is_prepare_running():
            QMessageBox.information(self, "等值线初稿", "单因素图仍在生成中。")
            return
        if self.is_contour_running():
            return
        self._set_generate_enabled(False)
        token = CancellationToken()
        worker = ContourDraftWorker(self._project, cancellation_token=token)
        self._contour_job.start(
            worker,
            terminal_signals=(worker.terminal,),
            result_connections=(
                (worker.completed, self._on_contour_completed),
                (worker.failed, self._on_contour_failed),
            ),
            cancel=token.cancel,
            target=self._project,
        )

    def _on_contour_completed(self, result: ContourDraftResult) -> None:
        target = self._contour_job.target
        if target is None or self._project is not target:
            return
        drafts = commit_contour_drafts(target, result)
        if not drafts:
            QMessageBox.information(
                self,
                "等值线初稿",
                "没有可提取的单因素网格。请先「批量生成单因素图」。",
            )
            return
        self.update_state(self._project.factor_map_tasks)
        self.contour_drafts_updated.emit()
        QMessageBox.information(
            self,
            "等值线初稿",
            f"已生成 {len(drafts)} 份等值线初稿并推送到编图。",
        )

    def _on_contour_failed(self, message: str) -> None:
        if self._contour_job.target is not self._project:
            return
        self.task_panel.summary_label.setText(f"等值线初稿失败：{message}")

    def _clear_contour_job(self) -> None:
        if not self.is_prepare_running():
            self._set_generate_enabled(True)

File: libs/ui_pages_data/src/qt/preparation_page.cpp
// UI-06 — preparation_page.cpp: PreparationPage Qt shell implementation.
//
// Port of paleo_workbench/ui/pages/preparation_page.py's page-level
// substance: the splitter composition, the prepare/contour worker
// lifecycle guards (generation guard, job-target guard, in-flight
// supersede on project switch) and the frozen summary strings from
// preparation_model. Panels and scientific backends stay injected seams —
// the closure installer binds the real ui_seqviz/ui_wellseis/
// ui_pages_mapedit panels and the ui_workers prepare/contour cores.

#include <pwb/ui_pages_data/qt/preparation_page.hpp>

#include <QLabel>
#include <QMessageBox>
#include <QPushButton>
#include <QSplitter>
#include <QVBoxLayout>
#include <QWidget>

#include <pwb/ui_pages_data/preparation_model.hpp>

namespace pwb::ui_pages_data::qt {
namespace {

// Honest not-installed panel surface — never a fake interactive page.
QLabel* placeholder_label(const QString& text, QWidget* parent) {
    auto* label = new QLabel(text, parent);
    label->setAlignment(Qt::AlignCenter);
    label->setObjectName(QStringLiteral("PreparationPanelPlaceholder"));
    return label;
}

}  // namespace

PreparationPage::PreparationPage(QWidget* parent) : QWidget(parent) {
    setObjectName(QStringLiteral("PreparationPage"));

    auto* root = new QVBoxLayout(this);
    root->setContentsMargins(0, 0, 0, 0);
    content_ = new QSplitter(Qt::Horizontal, this);
    content_->setObjectName(QStringLiteral("PreparationSplitter"));
    // Placeholders are replaced by set_task_panel / set_well_table_panel /
    // set_preview_grid / set_boundary_panel at install time.
    content_->addWidget(
        placeholder_label(QStringLiteral("制备面板未注入"), this));
    auto* center = new QSplitter(Qt::Vertical, content_);
    center->addWidget(
        placeholder_label(QStringLiteral("预览网格未注入"), this));
    center->addWidget(
        placeholder_label(QStringLiteral("井表面板未注入"), this));
    center->setSizes({240, 360});
    content_->addWidget(center);
    content_->addWidget(
        placeholder_label(QStringLiteral("边界面板未注入"), this));
    content_->setSizes({300, 620, 280});
    root->addWidget(content_, 1);
}

// -- panel seams ---------------------------------------------------------------

void PreparationPage::set_task_panel(FactorTaskPanelApi* panel) {
    if (panel == task_panel_) return;
    if (task_panel_ != nullptr) {
        disconnect(task_panel_, nullptr, this, nullptr);
    }
    task_panel_ = panel;
    if (task_panel_ != nullptr) {
        content_->replaceWidget(0, task_panel_);
        connect(task_panel_, &FactorTaskPanelApi::generate_requested, this,
                &PreparationPage::on_generate_requested);
        connect(task_panel_, &FactorTaskPanelApi::contour_draft_requested,
                this, &PreparationPage::on_contour_draft_requested);
        if (task_panel_->generate_btn() != nullptr &&
            task_panel_->contour_draft_btn() != nullptr) {
            QWidget::setTabOrder(task_panel_->generate_btn(),
                                 task_panel_->contour_draft_btn());
        }
        task_panel_->setVisible(true);
    }
}

void PreparationPage::set_well_table_panel(WellTablePanelApi* panel) {
    if (panel == well_table_panel_) return;
    if (well_table_panel_ != nullptr) {
        disconnect(well_table_panel_, nullptr, this, nullptr);
    }
    well_table_panel_ = panel;
    if (well_table_panel_ != nullptr) {
        // Center splitter: preview grid (index 0) above the well table.
        auto* center = qobject_cast<QSplitter*>(content_->widget(1));
        if (center != nullptr) center->replaceWidget(1, well_table_panel_);
        if (well_table_panel_->run_qc_btn() != nullptr) {
            connect(well_table_panel_->run_qc_btn(), &QPushButton::clicked,
                    this, &PreparationPage::on_run_well_qc);
        }
        well_table_panel_->setVisible(true);
    }
}

void PreparationPage::set_preview_grid(FactorPreviewGridApi* panel) {
    if (panel == preview_grid_) return;
    preview_grid_ = panel;
    if (preview_grid_ != nullptr) {
        auto* center = qobject_cast<QSplitter*>(content_->widget(1));
        if (center != nullptr) center->replaceWidget(0, preview_grid_);
        preview_grid_->setVisible(true);
    }
}

void PreparationPage::set_boundary_panel(QWidget* panel) {
    if (panel == boundary_panel_) return;
    boundary_panel_ = panel;
    if (boundary_panel_ != nullptr) {
        content_->replaceWidget(2, boundary_panel_);
        boundary_panel_->setVisible(true);
    }
}

// -- domain seams ---------------------------------------------------------------

void PreparationPage::set_generation_fns(std::function<int()> next_fn,
                                          std::function<int()> current_fn) {
    next_generation_fn_ = std::move(next_fn);
    current_generation_fn_ = std::move(current_fn);
}

void PreparationPage::set_snapshot_task_count_fn(
    std::function<int(void*, const std::string&, int)> fn) {
    snapshot_task_count_fn_ = std::move(fn);
}

void PreparationPage::set_prepare_worker_fn(
    std::function<void(void*, const std::string&, int,
                       std::function<void(const PrepareProgressView&)>,
                       std::function<void(const PrepareResultView&)>,
                       std::function<void(const QString&)>,
                       std::function<void()>)>
        fn) {
    prepare_worker_fn_ = std::move(fn);
}

void PreparationPage::set_commit_prepare_fn(
    std::function<int(void*, const PrepareResultView&, int)> fn) {
    commit_prepare_fn_ = std::move(fn);
}

void PreparationPage::set_contour_worker_fn(
    std::function<void(void*, std::function<void(void*)>,
                       std::function<void(const QString&)>)>
        fn) {
    contour_worker_fn_ = std::move(fn);
}

void PreparationPage::set_commit_contour_fn(
    std::function<int(void*, void*)> fn) {
    commit_contour_fn_ = std::move(fn);
}

void PreparationPage::set_display_well_table_fn(
    std::function<void*(void*, const domain::Json&)> fn) {
    display_well_table_fn_ = std::move(fn);
}

void PreparationPage::set_factor_map_tasks_fn(
    std::function<domain::Json(void*)> fn) {
    factor_map_tasks_fn_ = std::move(fn);
}

void PreparationPage::set_run_well_qc_fn(
    std::function<bool(void*, std::string*, std::string*)> fn) {
    run_well_qc_fn_ = std::move(fn);
}

// -- project / state ------------------------------------------------------------

void PreparationPage::set_project(void* project) {
    // Project switch supersedes any in-flight prepare generation.
    if (project != project_ && is_prepare_running()) {
        if (next_generation_fn_) {
            prepare_generation_ = next_generation_fn_();
        }
        if (prepare_job_.cancel) {
            prepare_job_.cancel();
        }
    }
    project_ = project;
    refresh_well_table_view();
}

void PreparationPage::update_state(const domain::Json& tasks) {
    tasks_ = tasks.is_array() ? tasks : domain::Json::array();
    if (task_panel_ != nullptr) {
        task_panel_->update_state(tasks_);
    }
    if (preview_grid_ != nullptr) {
        preview_grid_->update_state(tasks_);
    }
    refresh_well_table_view();
}

bool PreparationPage::is_prepare_running() const {
    return prepare_job_.is_running != nullptr && prepare_job_.is_running();
}

bool PreparationPage::is_contour_running() const {
    return contour_job_.is_running != nullptr && contour_job_.is_running();
}

bool PreparationPage::shutdown_workers(int wait_ms) {
    // A seam with no bound job is trivially joined (Python
    // OwnedWorkerJob.shutdown parity: nothing to join → True).
    const bool contour_joined = contour_job_.shutdown == nullptr ||
                                contour_job_.shutdown(wait_ms);
    const bool prepare_joined = prepare_job_.shutdown == nullptr ||
                                prepare_job_.shutdown(wait_ms);
    set_generate_enabled(true);
    return contour_joined && prepare_joined;
}

// -- internal helpers ------------------------------------------------------------

void PreparationPage::set_generate_enabled(bool enabled) {
    if (task_panel_ != nullptr) {
        if (auto* btn = task_panel_->generate_btn(); btn != nullptr) {
            btn->setEnabled(enabled);
        }
        if (auto* btn = task_panel_->contour_draft_btn(); btn != nullptr) {
            btn->setEnabled(enabled);
        }
    }
    if (well_table_panel_ != nullptr) {
        if (auto* btn = well_table_panel_->run_qc_btn(); btn != nullptr) {
            btn->setEnabled(enabled);
        }
    }
}

void PreparationPage::refresh_well_table_view() {
    if (well_table_panel_ == nullptr) return;
    void* table = nullptr;
    if (project_ != nullptr && display_well_table_fn_) {
        table = display_well_table_fn_(project_, tasks_);
    }
    well_table_panel_->update_from_well_table(table);
}

int PreparationPage::current_generation() const {
    return current_generation_fn_ ? current_generation_fn_()
                                  : prepare_generation_;
}

void* PreparationPage::prepare_job_target() const {
    return prepare_job_.target != nullptr ? prepare_job_.target() : nullptr;
}

void* PreparationPage::contour_job_target() const {
    return contour_job_.target != nullptr ? contour_job_.target() : nullptr;
}

// -- well QC -----------------------------------------------------------------------

void PreparationPage::on_run_well_qc() {
    if (project_ == nullptr) {
        QMessageBox::information(this, QStringLiteral("井点 QC"),
                                 QStringLiteral("请先打开或绑定工程。"));
        return;
    }
    if (!run_well_qc_fn_) {
        QMessageBox::warning(this, QStringLiteral("井点 QC 失败"),
                             QStringLiteral("QC 服务未接入"));
        return;
    }
    std::string info;
    std::string error;
    if (run_well_qc_fn_(project_, &info, &error)) {
        QMessageBox::information(this, QStringLiteral("井点 QC"),
                                 QString::fromStdString(info));
    } else {
        QMessageBox::warning(this, QStringLiteral("井点 QC 失败"),
                             QString::fromStdString(error));
    }
}

// -- prepare flow --------------------------------------------------------------------

void PreparationPage::on_generate_requested(const QString& method_q) {
    std::string method = method_q.toStdString();
    if (method.empty() && task_panel_ != nullptr) {
        method = task_panel_->selected_method().toStdString();
    }
    if (project_ == nullptr) {
        emit generate_requested(QString::fromStdString(method));
        return;
    }
    if (is_prepare_running()) {
        QMessageBox::information(this, QStringLiteral("单因素图"),
                                 QStringLiteral("正在生成中，请稍候…"));
        return;
    }
    start_prepare_worker(method);
}

void PreparationPage::start_prepare_worker(const std::string& method) {
    if (!prepare_worker_fn_) {
        return;  // no backend bound — stay inert, never a fake success
    }
    set_generate_enabled(false);
    prepare_generation_ =
        next_generation_fn_ ? next_generation_fn_() : prepare_generation_ + 1;
    const int generation = prepare_generation_;
    // Snapshot on the host thread so scientific inputs match the
    // Stage-4 fingerprints (Python parity).
    const int task_count = snapshot_task_count_fn_
                               ? snapshot_task_count_fn_(project_, method,
                                                         generation)
                               : 0;
    if (task_panel_ != nullptr && task_panel_->summary_label() != nullptr) {
        task_panel_->summary_label()
            ->setText(QString::fromStdString(
                prepare_start_label(task_count, generation)));
    }

    auto progress = [this](const PrepareProgressView& update) {
        on_prepare_progress(update);
    };
    auto completed = [this](const PrepareResultView& result) {
        on_prepare_completed(result);
    };
    auto failed = [this](const QString& message) {
        on_prepare_failed(message);
    };
    auto cancelled = [this]() { on_prepare_cancelled(); };
    // The host runs the worker off-thread and invokes exactly one terminal
    // callback on the UI thread (header contract).
    prepare_worker_fn_(project_, method, generation, progress, completed,
                       failed, cancelled);
}

void PreparationPage::clear_prepare_job() {
    if (!is_contour_running()) {
        set_generate_enabled(true);
    }
}

void PreparationPage::on_prepare_progress(const PrepareProgressView& update) {
    if (update.generation != current_generation()) return;
    const void* target = prepare_job_target();
    if (target != nullptr && target != project_) return;
    if (task_panel_ == nullptr || task_panel_->summary_label() == nullptr) {
        return;
    }
    task_panel_->summary_label()->setText(QString::fromStdString(
        prepare_progress_label(update.clean, update.dirty, update.completed,
                               update.total_tasks,
                               !update.message.empty() ? update.message
                                                       : update.phase)));
}

void PreparationPage::on_prepare_completed(const PrepareResultView& result) {
    if (project_ == nullptr) return;
    const void* target = prepare_job_target();
    if (target != nullptr && target != project_) return;
    if (result.generation != current_generation()) {
        // Superseded by a newer prepare request or project switch.
        return;
    }
    const int discarded =
        commit_prepare_fn_ ? commit_prepare_fn_(project_, result,
                                                current_generation())
                           : 0;
    if (factor_map_tasks_fn_) {
        update_state(factor_map_tasks_fn_(project_));
    }
    emit factor_maps_updated();
    if (task_panel_ != nullptr && task_panel_->summary_label() != nullptr) {
        int complete = 0;
        for (const auto& task : tasks_) {
            const auto status = task.find("status");
            if (status != task.end() && status->is_string() &&
                status->get<std::string>() == "complete") {
                ++complete;
            }
        }
        const int total = tasks_.is_array() ? int(tasks_.size()) : 0;
        task_panel_->summary_label()->setText(QString::fromStdString(
            prepare_done_label(complete, total, result.clean_count,
                               result.executed_count, discarded)));
    }
    clear_prepare_job();
}

void PreparationPage::on_prepare_failed(const QString& message) {
    const void* target = prepare_job_target();
    if (target != nullptr && target != project_) return;
    if (task_panel_ != nullptr && task_panel_->summary_label() != nullptr) {
        task_panel_->summary_label()->setText(QString::fromStdString(
            prepare_failed_label(message.toStdString())));
    }
    clear_prepare_job();
}

void PreparationPage::on_prepare_cancelled() {
    const void* target = prepare_job_target();
    if (target != nullptr && target != project_) return;
    if (task_panel_ != nullptr && task_panel_->summary_label() != nullptr) {
        task_panel_->summary_label()->setText(
            QString::fromStdString(std::string(kPrepareCancelledLabel)));
    }
    clear_prepare_job();
}

// -- contour flow ----------------------------------------------------------------------

void PreparationPage::on_contour_draft_requested() {
    if (project_ == nullptr) {
        QMessageBox::information(this, QStringLiteral("等值线初稿"),
                                 QStringLiteral("请先打开或绑定工程。"));
        return;
    }
    if (is_prepare_running()) {
        QMessageBox::information(this, QStringLiteral("等值线初稿"),
                                 QStringLiteral("单因素图仍在生成中。"));
        return;
    }
    if (is_contour_running()) return;
    if (!contour_worker_fn_) {
        return;  // no backend bound — stay inert
    }
    set_generate_enabled(false);
    auto completed = [this](void* result) { on_contour_completed(result); };
    auto failed = [this](const QString& message) {
        on_contour_failed(message);
    };
    contour_worker_fn_(project_, completed, failed);
}

void PreparationPage::clear_contour_job() {
    if (!is_prepare_running()) {
        set_generate_enabled(true);
    }
}

void PreparationPage::on_contour_completed(void* result) {
    if (project_ == nullptr) return;
    const void* target = contour_job_target();
    if (target != nullptr && target != project_) return;
    const int drafts = commit_contour_fn_
                           ? commit_contour_fn_(project_, result)
                           : 0;
    if (task_panel_ == nullptr || task_panel_->summary_label() == nullptr) {
        return;
    }
    if (drafts == 0) {
        // Async completions must not open a modal dialog (Python #897).
        task_panel_->summary_label()->setText(QString::fromStdString(
            contour_empty_label()));
        return;
    }
    if (factor_map_tasks_fn_) {
        update_state(factor_map_tasks_fn_(project_));
    }
    emit contour_drafts_updated();
    task_panel_->summary_label()->setText(
        QString::fromStdString(contour_done_label(drafts)));
    clear_contour_job();
}

void PreparationPage::on_contour_failed(const QString& message) {
    const void* target = contour_job_target();
    if (target != nullptr && target != project_) return;
    if (task_panel_ != nullptr && task_panel_->summary_label() != nullptr) {
        task_panel_->summary_label()->setText(QString::fromStdString(
            contour_failed_label(message.toStdString())));
    }
    clear_contour_job();
}

}  // namespace pwb::ui_pages_data::qt
